src/experimental/collection_new_iterable_model/shared/iterator/RedBlackTreeIterator.hpp

```diff
@@ -0,0 +1,397 @@
+            }
+        }
+
+        template < typename __ElementType >     // NOLINT(bugprone-reserved-identifier)
+        class AbstractTreeIterator : public meta :: BidirectionalIterator {
+            template <
+                    typename __TElementType,                                                                // NOLINT(bugprone-reserved-identifier)
+                    typename __KeyType,                                                                     // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: ExtractorFunction < __TElementType, __KeyType > __keyExtractor,       // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: ComparisonFunction < __TElementType > __keyLowerComparator,           // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: ComparisonFunction < __TElementType > __keyEqualsComparator,          // NOLINT(bugprone-reserved-identifier)
+                    cds :: utility :: DestructorFunction < __TElementType > __nodeDestructor                // NOLINT(bugprone-reserved-identifier)
+            > friend class cds :: experimental :: __hidden :: __impl :: __RedBlackTree;                     // NOLINT(bugprone-reserved-identifier)
+
+
+            protected:
+                using RBTreeNode = cds :: __hidden :: __impl :: __RedBlackTreeNode < __ElementType >;
+
+            private:
+                RBTreeNode *         _pRoot          { 0ULL };
+
+            private:
+                RBTreeNode *         _pCurrentNode  { nullptr };
+
+            protected:
+                AbstractTreeIterator() noexcept;
+
+            protected:
+                AbstractTreeIterator(
+                        RBTreeNode * pRoot,
+                        RBTreeNode * pCurrentNode
+                ) noexcept;
+
+            protected:
+                AbstractTreeIterator(
+                        AbstractTreeIterator const & iterator
+                ) noexcept;
+
+            protected:
+                AbstractTreeIterator(
+                        AbstractTreeIterator && iterator
+                ) noexcept;
+
+            protected:
+
+            public:
+                auto operator * () const noexcept -> __ElementType &;
+
+            public:
+                auto operator -> () const noexcept -> __ElementType *;
+
+            public:
+                auto operator == (
+                        AbstractTreeIterator const & iterator
+                ) const noexcept -> bool;
+
+            public:
+                auto operator != (
+                        AbstractTreeIterator const & iterator
+                ) const noexcept -> bool;
+
+            public:
+                __CDS_Explicit operator bool () const noexcept;
+
+            protected:
+                auto successor () noexcept -> AbstractTreeIterator &;
```

Comment:
can we use previous / next instead? are successor / predecessor tree-specific terms?